huaweicloud/config/auth.go

```diff
@@ -0,0 +1,374 @@
+	if c.AgencyDomainName != "" && c.AgencyName != "" {
+		pao = golangsdk.AuthOptions{
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+			DelegatedProject: c.DelegatedProject,
+		}
+
+		dao = golangsdk.AuthOptions{
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+		}
+	} else {
+		pao = golangsdk.AuthOptions{
+			DomainID:   c.DomainID,
+			DomainName: c.DomainName,
+			TenantID:   c.TenantID,
+			TenantName: c.TenantName,
+		}
+
+		dao = golangsdk.AuthOptions{
+			DomainID:   c.DomainID,
+			DomainName: c.DomainName,
+		}
+	}
+
+	for _, ao := range []*golangsdk.AuthOptions{&pao, &dao} {
+		ao.IdentityEndpoint = c.IdentityEndpoint
+		ao.TokenID = c.Token
+
+	}
+	return genClients(c, pao, dao)
+}
+
+func buildClientByAKSK(c *Config) error {
+	var pao, dao golangsdk.AKSKAuthOptions
+
+	if c.AgencyDomainName != "" && c.AgencyName != "" {
+		pao = golangsdk.AKSKAuthOptions{
+			DomainID:         c.DomainID,
+			Domain:           c.DomainName,
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+			DelegatedProject: c.DelegatedProject,
+		}
+
+		dao = golangsdk.AKSKAuthOptions{
+			DomainID:         c.DomainID,
+			Domain:           c.DomainName,
+			AgencyName:       c.AgencyName,
+			AgencyDomainName: c.AgencyDomainName,
+		}
+	} else {
+		pao = golangsdk.AKSKAuthOptions{
+			BssDomainID: c.DomainID,
+			BssDomain:   c.DomainName,
+			ProjectName: c.TenantName,
+			ProjectId:   c.TenantID,
+		}
+
+		dao = golangsdk.AKSKAuthOptions{
+			DomainID: c.DomainID,
+			Domain:   c.DomainName,
+		}
+	}
+
+	for _, ao := range []*golangsdk.AKSKAuthOptions{&pao, &dao} {
+		ao.IdentityEndpoint = c.IdentityEndpoint
+		ao.AccessKey = c.AccessKey
+		ao.SecretKey = c.SecretKey
+		if c.Region != "" {
+			ao.Region = c.Region
+		}
+		if c.SecurityToken != "" {
+			ao.SecurityToken = c.SecurityToken
+			ao.WithUserCatalog = true
+		}
+	}
+	return genClients(c, pao, dao)
+}
+
+func buildClientByConfig(c *Config) error {
+	profilePath, err := homedir.Expand(c.SharedConfigFile)
+	if err != nil {
+		return err
+	}
+
+	current := c.Profile
+	var providerConfig Profile
+	_, err = os.Stat(profilePath)
+	if !os.IsNotExist(err) {
+		data, err := ioutil.ReadFile(profilePath)
+		if err != nil {
+			return fmt.Errorf("Err reading from shared config file: %s", err)
+		}
+		sharedConfig := SharedConfig{}
+		err = json.Unmarshal(data, &sharedConfig)
+		if err != nil {
+			return err
+		}
+
+		// fetch current from shared config if not specified with provider
+		if current == "" {
+			current = sharedConfig.Current
+		}
+
+		// fetch the current profile config
+		for _, v := range sharedConfig.Profiles {
+			if current == v.Name {
+				providerConfig = v
+				break
+			}
+		}
+		if (providerConfig == Profile{}) {
+			return fmt.Errorf("Error finding profile %s from shared config file", current)
+		}
+	} else {
+		return fmt.Errorf("The specified shared config file %s does not exist", profilePath)
+	}
```

Comment:
Suggest to put the `else` logic in the `if`, and extract the logic in the `if` (means isExist()).

Reply:
done